examples/tests/app_loader/main.c

```diff
@@ -0,0 +1,298 @@
+  0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+  0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+  0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0};
+
+
+/******************************************************************************************************
+* Callback functions
+*
+* 1. Callback to let us know when the capsule is done writing data to flash
+* 2. Set button callback to initiate the dynamic app load process on pressing button 1 (on nrf52840dk)
+*
+******************************************************************************************************/
+
+// static void nop_callback(int a __attribute__((unused)), int b __attribute__((unused)), int c __attribute__((unused)), void *d __attribute__((unused))) {}
+
+static void app_write_done_callback(__attribute__((unused)) int   length,
+                                    __attribute__((unused)) int   arg1,
+                                    __attribute__((unused)) int   arg2,
+                                    __attribute__((unused)) void *ud)
+{
+  done = true;
+}
+
+static void button_callback(int                            btn_num,
+                            int                            val,
+                            __attribute__ ((unused)) int   arg2,
+                            __attribute__ ((unused)) void *ud)
+{
+  // Callback for button presses.
+  // btn_num: The index of the button associated with the callback
+  // val: 1 if pressed, 0 if depressed
+  if ((btn_num == 0) && (val == 1)) {
+    printf("[Event] Button Pressed!\n");
+
+    int ret0;
+    ret0 = app_loader_exists();
+    if (ret0 == RETURNCODE_SUCCESS) {
+      printf("[Success] Dynamic Apploader driver exists.\n");
+    }
+
+    int ret  = 0;
+    int ret1 = 0;
+    int ret2 = 0;
+
+    app_size = sizeof(app_binary);
+    ret      = app_loader_command_setup(app_size); // asks the capsule to set up for app flash
+    if (ret != RETURNCODE_SUCCESS) {
+      printf("[Error] Setup Failed: %d.\n", ret);
+    }
+
+    if (ret == RETURNCODE_SUCCESS) {
+      printf("[Success] Setup successful. Attempting to write app to flash now.\n");
+      ret1 = write_app();                           // writes app data to flash
+      if (ret1 != RETURNCODE_SUCCESS) {
+        printf("[Error] App flash write unsuccessful: %d.\n", ret1);
+      }
+    }
+
+    if (ret1 == RETURNCODE_SUCCESS) {
+      printf("[Success] App flashed successfully. Attempting to create a process now.\n");
+      ret2 = app_loader_command_load();             // request to load app
+      if (ret2 != RETURNCODE_SUCCESS) {
+        printf("[Error] Process creation failed: %d.\n", ret2);
+      } else {
+        printf("[Success] Process created successfully.\n");
+      }
+    }
+    app_size = 0; // reset app_size
+  }
+
+  if ((btn_num == 1) && (val == 1)) {
+    printf("Button 2 pressed!\n");
+  }
```

Comment:
Why not just support any button? It's not always obvious which button is button 1 on a board

Reply:
In this case, I used button 1 to load the app, and button 2 to print the statement to ensure I could get the buttons to work independently. I am removing it, but it can be added back anytime.